### PR TITLE
static turn instructions

### PR DESCRIPTION
**Pull request by MoKob**
For this branch to be merged, the following needs to happen:

 - [x] do not expose the internal conflict resolution
 - [x] do not expose `new name` on near identical names
 - [x] add road classes (e.g. [primary vs residential](http://www.openstreetmap.org/edit#map=18/41.69060/-120.37531)) to fork detection
 - [x] clean-up response to only return valid turn candidates
 
~~- [ ] add cucumber tests for static turn instructions~~
~~- [x] motorways~~
~~- [ ] fork~~
~~- [ ] end of road~~
~~- [ ] new name~~
~~- [ ] suppressed~~
~~- [ ] continue~~
~~- [ ] merge~~
~~- [ ] ramp~~
 - [x] test roundabouts
      - [x] enter
      - [x] immediately exit
      - [x] enter and exit
      - [x] only exit
 - [x] review, I am looking at you @daniel-j-h

### CODE REVIEW

File: include/extractor/guidance/turn_instruction.hpp [hunk 1]
```diff
@@ -43,8 +43,14 @@ enum TurnType // at the moment we can support 32 turn types, without increasing
     NewName,                // no turn, but name changes
```
Comment (TheMarex) [inline include/extractor/guidance/turn_instruction.hpp]: Can we merge `NewName` and `Notification`? `Notification` just tells the user that a street attribute (like name or travel mode) changed. This would be forward compatible if we add more attributes for which we might want to emit a `Notification.

File: include/extractor/guidance/turn_instruction.hpp [hunk 2]
```diff
@@ -43,8 +43,14 @@ enum TurnType // at the moment we can support 32 turn types, without increasing
...
     Continue,               // remain on a street
     Turn,                   // basic turn
+    FirstTurn,              // First of x turns
+    SecondTurn,             // Second of x turns
+    ThirdTurn,              // Third of x turns
     Merge,                  // merge onto a street
     Ramp,                   // special turn (highway ramp exits)
+    FirstRamp,              // numbered ramp
```
Comment (TheMarex) [inline include/extractor/guidance/turn_instruction.hpp]: Do you see any way of moving this additional information outside the turn type? I mean "only" handling four ramps at the same node is fine with me, but how will this evolve once we also emit intersection shapes?

**Comment by TheMarex** [comment]
:warning: rebased on the new `rewrite/new-api`.

**Comment by TheMarex** [comment]
:warning: rebased on `rewrite/new-api`.

File: include/extractor/guidance/turn_instruction.hpp (hunk 1 above)
Comment (MoKob) [inline include/extractor/guidance/turn_instruction.hpp]: Right Now Notification is Unused due to data not being exposed. I wouldn't worry about that one right now. The moment we get more data into the system, we have to open a discussion of what to expose.

File: include/extractor/guidance/toolkit.hpp [hunk 3]
```diff
@@ -12,9 +12,12 @@
 #include "extractor/guidance/classification_data.hpp"
 #include "extractor/guidance/turn_instruction.hpp"
 
+#include <algorithm>
 #include <map>
 #include <cmath>
 #include <cstdint>
+#include <string>
+#include <iostream>
```
Comment (daniel-j-h) [inline include/extractor/guidance/toolkit.hpp]: can be removed?

File: include/extractor/guidance/toolkit.hpp [hunk 4]
```diff
@@ -315,6 +323,84 @@ inline bool isLowPriorityRoadClass(const FunctionalRoadClass road_class)
...
+
+    if ((second + 1) % detail::num_direction_modifiers == first)
+        return false;
+
+    return true;
+}
+
+inline bool requiresNameAnnounced(const std::string &from, const std::string &to)
+{
+    // FIXME, handle in profile to begin with?
```
Comment (daniel-j-h) [inline include/extractor/guidance/toolkit.hpp]: yes we should ticket that; also the "Weg" handling below is ugly here. Maybe for post-v5?

File: include/extractor/guidance/toolkit.hpp [hunk 5]
```diff
@@ -315,6 +323,84 @@ inline bool isLowPriorityRoadClass(const FunctionalRoadClass road_class)
...
+            to_ref.find(from_ref) != std::string::npos)
+            return false;
+    }
+    return true;
+}
+
+inline int getPriority( const FunctionalRoadClass road_class )
+{
+    const constexpr int road_priority[] = {10, 0, 10, 2, 10, 4, 10, 6, 10, 8, 10, 11, 10, 12, 10, 14};
+    return road_priority[static_cast<int>(road_class)];
```
Comment (daniel-j-h) [inline include/extractor/guidance/toolkit.hpp]: Where did you get those priorities from? I can see them coming from the functional classes, but it's hard to follow here.

File: include/extractor/guidance/toolkit.hpp [hunk 6]
```diff
@@ -315,6 +323,84 @@ inline bool isLowPriorityRoadClass(const FunctionalRoadClass road_class)
...
+}
+
+inline bool canBeSeenAsFork(const FunctionalRoadClass first, const FunctionalRoadClass second)
+{
+    // forks require similar road categories
+    return std::abs(getPriority(first) - getPriority(second)) <= 1;
```
Comment (daniel-j-h) [inline include/extractor/guidance/toolkit.hpp]: Why? Care to elaborate how this influences decisions? I also can't find users in this pull request. Is this for future use?

File: include/extractor/guidance/turn_analysis.hpp [hunk 7]
```diff
@@ -20,42 +23,61 @@ namespace extractor
...
+// For the turn analysis, we require a full list of all connected roads to determine the outcome.
+// Invalid turns can influence the perceived angles
+//
+// aaa(2)aa
+//          a - bbbbb
+// aaa(1)aa
+//
+// will not be perceived as a turn from (1) -> b, and as a U-turn from (1) -> (2).
+// In addition, they can influence whether a turn is obvious or not.
+struct ConnectedRoad : public TurnOperation
```
Comment (daniel-j-h) [inline include/extractor/guidance/turn_analysis.hpp]: Is this semantically correct? A connected road is a turn operation?

File: include/extractor/guidance/turn_analysis.hpp [hunk 8]
```diff
@@ -82,10 +106,9 @@ class TurnAnalysis
     //
     // The treatment results in a straight turn angle of 180º rather than a turn angle of approx
     // 160
-    std::vector<TurnCandidate>
-    mergeSegregatedRoads(const NodeID from_node,
-                         const EdgeID via_eid,
-                         std::vector<TurnCandidate> turn_candidates) const;
+    std::vector<ConnectedRoad> mergeSegregatedRoads(const NodeID from_node,
+                                                    const EdgeID via_eid,
+                                                    std::vector<ConnectedRoad> intersection) const;
```
Comment (daniel-j-h) [inline include/extractor/guidance/turn_analysis.hpp]: fyi, intersection is passed by value, is this what you want to do here?

File: include/extractor/guidance/turn_analysis.hpp [hunk 9]
```diff
@@ -93,91 +116,69 @@ class TurnAnalysis
...
     // Produces instructions to enter/exit a roundabout or to stay on it.
     // Performs the distinction between roundabout and rotaries.
-    std::vector<TurnCandidate> handleRoundabouts(const NodeID from,
+    std::vector<ConnectedRoad> handleRoundabouts(const NodeID from,
                                                  const EdgeID via_edge,
                                                  const bool on_roundabout,
                                                  const bool can_enter_roundabout,
                                                  const bool can_exit_roundabout,
-                                                 std::vector<TurnCandidate> turn_candidates) const;
+                                                 std::vector<ConnectedRoad> intersection) const;
```
Comment (daniel-j-h) [inline include/extractor/guidance/turn_analysis.hpp]: same here ^

File: include/extractor/guidance/turn_instruction.hpp [hunk 10]
```diff
@@ -42,8 +42,16 @@ enum TurnType // at the moment we can support 32 turn types, without increasing
     NewName,                // no turn, but name changes
     Continue,               // remain on a street
     Turn,                   // basic turn
+    FirstTurn,              // First of x turns
+    SecondTurn,             // Second of x turns
+    ThirdTurn,              // Third of x turns
+    FourthTurn,             // Fourth of x turns
```
Comment (daniel-j-h) [inline include/extractor/guidance/turn_instruction.hpp]: per the zero-one-infinity rule: why do we handle 1-4 explicitly?

File: include/util/name_table.hpp [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+namespace util
+{
+class NameTable
+{
+  private:
+    //FIXME should this use shared memory
+    RangeTable<16, false> m_name_table;
+    ShM<char, false>::vector m_names_char_list;
+  public:
+    NameTable( const std::string &filename );
```
Comment (daniel-j-h) [inline include/util/name_table.hpp]: format

File: include/util/name_table.hpp [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+#include "util/shared_memory_vector_wrapper.hpp"
+#include "util/range_table.hpp"
+
+#include <string>
+
+namespace osrm
+{
+namespace util
+{
+class NameTable
```
Comment (daniel-j-h) [inline include/util/name_table.hpp]: what is a name table good for, what's its purpose? skimming the class says nothing about it. I assume for street names?

File: src/extractor/edge_based_graph_factory.cpp [hunk 13]
```diff
@@ -302,8 +302,8 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
     // Three nested loop look super-linear, but we are dealing with a (kind of)
     // linear number of turns only.
     util::Percent progress(m_node_based_graph->GetNumberOfNodes());
-    guidance::TurnAnalysis turn_analysis( *m_node_based_graph, m_node_info_list,
-                                       *m_restriction_map, m_barrier_nodes, m_compressed_edge_container );
+    guidance::TurnAnalysis turn_analysis(*m_node_based_graph, m_node_info_list, *m_restriction_map,
```
Comment (daniel-j-h) [inline src/extractor/edge_based_graph_factory.cpp]: Hm why is restriction map wrapped in a shared_ptr?

File: src/util/name_table.cpp [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+#include <boost/filesystem/fstream.hpp>
+
+namespace osrm
+{
+namespace util
+{
+
+NameTable::NameTable(const std::string &filename)
+{
+    boost::filesystem::ifstream name_stream(filename, std::ios::binary);
```
Comment (daniel-j-h) [inline src/util/name_table.cpp]: if (!name_stream) throw

File: src/util/name_table.cpp [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+    name_stream >> m_name_table;
+
+    unsigned number_of_chars = 0;
+    name_stream.read(reinterpret_cast<char *>(&number_of_chars), sizeof(number_of_chars));
+    BOOST_ASSERT_MSG(0 != number_of_chars, "name file broken");
+    m_names_char_list.resize(number_of_chars + 1); //+1 gives sentinel element
```
Comment (daniel-j-h) [inline src/util/name_table.cpp]: What is the sentinel needed for? the number_of_chars + 1 elements are default inserted, which 1/ allocated memory and 2/ initializes them to zero. Do you only care for the last zero?

Comment (MoKob) [inline src/util/name_table.cpp]: I honestly didn't think about it. I just copied it from the datafacade. The problem here, I require the names which have been serialized to disk and cannot be accessed later without reeding them back in. I cannot think of a reasonable way to have this in here, but @TheMarex reasonably said that some of these tests should be performed in pre-processing, if possible

File: include/extractor/guidance/toolkit.hpp (hunk 6 above)
Comment (MoKob) [inline include/extractor/guidance/toolkit.hpp]: They are used in the not shown `TurnAnalsyis.cpp`. It essentially decides that we do not discover a fork for a service road leaving a motorway.

File: include/extractor/guidance/turn_analysis.hpp (hunk 7 above)
Comment (MoKob) [inline include/extractor/guidance/turn_analysis.hpp]: Technically it contains a possible turn operation with a flag. We could set the Turn Operation as variable for that struct. Should not change much, but would elongate all accesses down the line.

File: include/extractor/guidance/turn_analysis.hpp (hunk 8 above)
Comment (MoKob) [inline include/extractor/guidance/turn_analysis.hpp]: I tried to follow your guidance for inplace modification + move in this pull request. So yes, it should be this way.

File: include/extractor/guidance/turn_instruction.hpp (hunk 10 above)
Comment (MoKob) [inline include/extractor/guidance/turn_instruction.hpp]: because this is the only thing that happens in not fully messed up data-modelling. 
Haven;t seen anything that has more than four turns on a single side and wasn't a data modelling problem. They will probably make room for something better down the line.

File: src/extractor/edge_based_graph_factory.cpp (hunk 13 above)
Comment (MoKob) [inline src/extractor/edge_based_graph_factory.cpp]: Good question. It has been there and I didn't wand to mix clean-up with new features too much.

**Comment by MoKob** [comment]
Moved the remaining tests here: https://github.com/Project-OSRM/osrm-backend/issues/2114

**Comment by TheMarex** [comment]
:warning: rebasing on `rewrite/new-api`.

File: include/util/name_table.hpp [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+{
+  private:
+    // FIXME should this use shared memory
+    RangeTable<16, false> m_name_table;
+    ShM<char, false>::vector m_names_char_list;
+
+  public:
+    NameTable(const std::string &filename);
+    std::string get_name_for_id(const unsigned name_id) const;
```
Comment (TheMarex) [inline include/util/name_table.hpp]: Needs to be CamelCase.

File: include/extractor/guidance/toolkit.hpp (hunk 4 above)
Comment (TheMarex) [inline include/extractor/guidance/toolkit.hpp]: Agreed. We can move the more complex merging out of this PR to more sophisticated methods. I'm going to simplify the code below and PR the diff to the current code.

**Comment by TheMarex** [comment]
Okay this is in good shape now. Good work @MoKob! Merging back to the rewrite branch.
